Add boundary condition support to Creators::Interval

nilsdeppe opened the pull request:
## Proposed changes

Adds support for specifying boundary conditions at runtime and adding them to the domain to the interval domain creator.

Some notes:
- In order to not break develop and not have one 15,000 line PR (not an exaggeration...) I have to support both having and not having boundary conditions specified in the systems. Once all the systems (both elliptic and evolution) are migrated over to supporting boundary conditions, we can clean up the code. This is noted in #2779 along with some related cleanups
- There are some "extra" or "unexpected" changes, such as the need to register boundary conditions in some of the elliptic test executables.
- Once folks have had a chance to look at the interface, I'll open up PRs for all the remaining domain creators. I have not added boundary condition support for all conceivable cases of option values to all domain creators because I'm not sure we will ever use some of them, and if we do, adding them isn't so difficult that whoever wants that configuration couldn't add the support.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add boundary condition support to Creators::Interval

nilsvu commented on src/Domain/Creators/Interval.hpp:
```diff
@@ -69,19 +71,56 @@ class Interval : public DomainCreator<1> {
+  };
 
-  using options =
+  using basic_options =
       tmpl::list<LowerBound, UpperBound, IsPeriodicIn, InitialRefinement,
                  InitialGridPoints, TimeDependence>;
 
+  template <typename Metavariables>
+  using options = tmpl::conditional_t<
+      domain::BoundaryConditions::has_boundary_conditions_base_v<
+          typename Metavariables::system>,
+      tmpl::push_back<
+          basic_options,
+          LowerBoundaryCondition<
+              domain::BoundaryConditions::get_boundary_conditions_base<
+                  typename Metavariables::system>>,
+          UpperBoundaryCondition<
+              domain::BoundaryConditions::get_boundary_conditions_base<
+                  typename Metavariables::system>>>,
+      basic_options>;
+
```
How about this pattern:

```cpp
template <typename Metavariables>
using options = tmpl::append<
    tmpl::list<LowerBound, UpperBound, IsPeriodicIn, InitialRefinement,
               InitialGridPoints, TimeDependence>,
    tmpl::conditional_t<
        domain::BoundaryConditions::has_boundary_conditions_base_v<
            typename Metavariables::system>,
        tmpl::list<LowerBoundaryCondition<
                       domain::BoundaryConditions::get_boundary_conditions_base<
                           typename Metavariables::system>>,
                   UpperBoundaryCondition<
                       domain::BoundaryConditions::get_boundary_conditions_base<
                           typename Metavariables::system>>>,
        tmpl::list<>>>;
```

I find this easier to understand, and it's easier to extend the list with more conditionals.

nilsdeppe replied:
I find this much harder to understand. Since this is getting removed after all the systems have boundary conditions anyway, I don't see it as being worth the effort to change in every domain creator.

If we want to support domains without systems we should probably completely rethink afterwards, since this approach is admittedly tedious but I couldn't think of a better way.

nilsvu commented on tests/Unit/Domain/Creators/Test_AlignedLattice.cpp:
```diff
@@ -44,16 +46,18 @@ void test_aligned_blocks(
+      "  InitialLevels: [2]\n"
+      "  RefinedLevels: []\n"
+      "  RefinedGridPoints: []\n"
+      "  BlocksToExclude: []\n");
```
This looks like it belongs in the next PR

nilsdeppe replied:
No, factory creation needs to be able to create all domain creators, and `Interval` no longer supports `void` as a metavariable

nilsvu commented on tests/Unit/Domain/Creators/CMakeLists.txt:
```diff
@@ -26,5 +26,15 @@ add_test_library(
+  ${LIBRARY}
+  PRIVATE
+  Domain
+  DomainBoundaryConditionsHelpers
```
Add `DomainBoundaryConditions`

nilsvu commented on tests/Unit/Domain/Creators/Test_Interval.cpp:
```diff
@@ -83,80 +90,174 @@ void test_interval() {
+        std::vector<std::unordered_set<Direction<1>>>{{}}, std::tuple<>{}, {},
+        {});
+  }
+  { INFO("With boundary condition"); }
```
This looks like it wants something to do :)

nilsvu commented on tests/Unit/Domain/Creators/Test_Interval.cpp:
```diff
@@ -83,80 +90,174 @@ void test_interval() {
+  const std::vector<DirectionMap<1, BlockNeighbor<1>>> expected_neighbors{
+      {{Direction<1>::lower_xi(), {0, {}}},
+       {Direction<1>::upper_xi(), {0, {}}}}};
+  {{INFO("Interval factory time independent, no boundary condition");
```
Please clang-format this file

Also I think you can remove the outer scopes

nilsvu commented on tests/Unit/ParallelAlgorithms/LinearSolver/Schwarz/Test_SchwarzAlgorithm.yaml:
```diff
@@ -20,6 +20,15 @@ DomainCreator:
+      UpperBoundary:
+        TestBoundaryCondition:
+          Direction: upper-xi
+          BlockId: 0
```
I'm fine with making these changes, though you could get around that by specifying the test-system without boundary conditions. We may end up with a way of avoiding dummy boundary conditions for things like this or ExportCoordinates or ApplyObservers, so it might be easier to do what you did for ExportCoordinates and just add the empty system here. Up to you.

nilsdeppe replied:
I made the change anticipating you'd want to add boundary conditions into these tests afterward :)

nilsvu replied:
Ah I see. Would you mind going with the empty-system approach for now then? The linear solver tests won't use boundary conditions because they work with hard-coded matrices.

nilsdeppe replied:
sure!

nilsvu commented on tests/Unit/Domain/Creators/Test_RotatedIntervals.cpp:
```diff
@@ -109,15 +111,17 @@ void test_rotated_intervals_factory() {
+      "  UpperBound: [1.0]\n"
+      "  IsPeriodicIn: [True]\n"
+      "  InitialGridPoints: [[3,2]]\n"
+      "  InitialRefinement: [2]\n");
```
This also looks like it belongs in the next PR

nilsdeppe replied:
No, factory creation needs to be able to create all domain creators, and `Interval` no longer supports `void` as a metavariable

nilsvu reviewed (changes requested):
Please squash these directly

nilsdeppe commented:
Rebased and squashed. Thanks for reviewing :)

nilsdeppe commented:
Rebased and removed the boundary conditions from the elliptic tests. Thanks for the review :)

fmahebert commented on src/Domain/Creators/Interval.cpp:
```diff
@@ -49,12 +52,36 @@ Interval::Interval(
+      boundary_conditions_all_blocks{};
+  if (lower_boundary_condition_ != nullptr) {
+    if (upper_boundary_condition_ == nullptr)  {
+      ERROR(
```
This ERROR will trigger when only a lower boundary condition is specified. I suggest adding a matching ERROR for the opposite case, i.e. when only the upper boundary condition is specified.

fmahebert commented on src/Domain/Creators/Interval.cpp:
```diff
@@ -49,12 +52,36 @@ Interval::Interval(
+          "neither.");
+    }
+    if (is_periodic_in_x_[0]) {
+      ERROR(
```
Likewise, should check periodic isn't paired with only an upper boundary condition

fmahebert commented on src/Executables/ExportCoordinates/ExportCoordinates.hpp:
```diff
@@ -180,6 +180,9 @@ template <size_t Dim>
   static constexpr size_t volume_dim = Dim;
   static constexpr bool local_time_stepping = false;
+  // all the domain creators need a system now. Currently don't require boundary
+  // conditions, so can wait to add placeholders for those.
```
[minor] I slightly dislike "now", "currently", because it's not clear to a future reader what times are being referred to. I suggest simply saying: "A placeholder system for the domain creator."

fmahebert commented on src/Domain/Creators/Interval.cpp:
```diff
@@ -49,12 +52,36 @@ Interval::Interval(
+      1, std::unique_ptr<domain::BoundaryConditions::BoundaryCondition>>>
+      boundary_conditions_all_blocks{};
+  if (lower_boundary_condition_ != nullptr) {
+    if (upper_boundary_condition_ == nullptr)  {
```
Question to check my understanding of the interface: what does the user do in the input file to get a nullptr here? Is it by leaving the option blank, like this?
```
BoundaryConditions:
  LowerBoundary:
  UpperBoundary:
```

nilsdeppe replied:
This is an annoying one that I haven't settled yet... I think the best option is to remove the `IsPeriodicIn` stuff completely and add a system-agnostic `Periodic` condition that is checked.

fmahebert reviewed (commented):
Some thoughts from a quick pass

kidder commented on tests/Unit/Domain/Creators/Test_Interval.cpp:
```diff
@@ -83,80 +90,198 @@ void test_interval() {
+            CoordinateMaps::TimeDependent::Translation{"TranslationX"}),
+        expected_boundary_conditions);
   }
 }
```
test the new ERRORs you introduced

nilsdeppe committed:
Add Periodic boundary condition class

This class is used to mark a system-specific boundary condition as representing
a periodic boundary condition. The system-specific boundary condition needs to
inherit from the Periodic class as well as the system-specific boundary
condition base class. Domain creators then use the is_periodic function to check
if a system-specific boundary condition is marked as being periodic.

nilsdeppe committed:
Clang-tidy fixes

nilsdeppe commented:
I have pushed commits (won't pass CI until after Blaze 3.8 is merged) that do the following:
- a new commit that adds a `Periodic` base class into the `Domain` that any system-specific periodic boundary conditions must inherit from to mark them as periodic. I intentionally did not do a virtual function because somebody will inevitably copy-pasta and create a "periodic" condition that isn't actually periodic. The design in this PR makes that rather difficult, and the tests will catch it.
- added a fixup commit that pretty much completely reworks `Interval` to support periodic boundary conditions with the new boundary condition interface. I've tested this in evolutions with scalar wave, and it behaves as expected. I'll propagate to the other domain creators once the design is settled, but I'm happy with this.

nilsvu commented on src/Domain/Creators/Interval.hpp:
```diff
@@ -76,51 +76,61 @@ class Interval : public DomainCreator<1> {
     static constexpr Options::String help =
         "Options for the boundary condition applied at the lower boundary.";
     using type = std::unique_ptr<BoundaryConditionsBase>;
-    using group = BoundaryConditions;
```
Why did you remove this group? I liked the `BoundaryConditions:` section in the input file.

nilsdeppe replied:
I had switched to using `Alternatives` at an intermediate stage. I agree, it's nice to have the group. I'll restore that. Thanks!

nilsvu commented on src/Domain/Creators/Interval.hpp:
```diff
@@ -76,51 +76,61 @@ class Interval : public DomainCreator<1> {
+          options_boundary_conditions<typename Metavariables::system>,
+          options_periodic>,
+      // Need TimeDependence at the end to not break the already broken python
+      // bindings. This is so that the constructors can default them to nullptr.
```
`git apply` this, then you can pass `nullptr`s to the constructor in the order you want:

```diff
diff --git a/src/Domain/Creators/Python/Interval.cpp b/src/Domain/Creators/Python/Interval.cpp
index e3027692f..0c7a80af4 100644
--- a/src/Domain/Creators/Python/Interval.cpp
+++ b/src/Domain/Creators/Python/Interval.cpp
@@ -17,9 +17,18 @@ namespace creators {
 namespace py_bindings {
 void bind_interval(py::module& m) {  // NOLINT
   py::class_<Interval, DomainCreator<1>>(m, "Interval")
-      .def(py::init<std::array<double, 1>, std::array<double, 1>,
-                    std::array<bool, 1>, std::array<size_t, 1>,
-                    std::array<size_t, 1>>(),
+      .def(py::init(
+               [](std::array<double, 1> lower_x, std::array<double, 1> upper_x,
+                  std::array<bool, 1> is_periodic_in_x,
+                  std::array<size_t, 1> initial_refinement_level_x,
+                  std::array<size_t, 1> initial_number_of_grid_points_in_x) {
+                 return Interval{std::move(lower_x),
+                                 std::move(upper_x),
+                                 std::move(is_periodic_in_x),
+                                 std::move(initial_refinement_level_x),
+                                 std::move(initial_number_of_grid_points_in_x),
+                                 nullptr};
+               }),
            py::arg("lower_x"), py::arg("upper_x"), py::arg("is_periodic_in_x"),
            py::arg("initial_refinement_level_x"),
            py::arg("initial_number_of_grid_points_in_x"));
```

nilsdeppe commented:
- rebased on develop
- added new fixup that restores the group, and removed the ` = nullptr` argument (though I may need to undo this if we want to support compile-time no time dependence)

fmahebert commented on src/Domain/BoundaryConditions/Periodic.hpp:
```diff
@@ -0,0 +1,33 @@
+///
+/// Periodic boundary conditions shouldn't require any implementation outside of
+/// a check in the domain creator using the `is_periodic()` function to
+/// determine what boundaries are periodic.
```
Consider appending a sentence like "Across each matching pair of periodic boundary conditions, the domain creator should specify that the DG elements are neighbors of each other."

My thinking is to make sure users understand the domain creator has to do something with the information... you can't just put a periodic BC in the input file and expect magic.

fmahebert commented on src/Domain/Creators/Interval.hpp:
```diff
@@ -75,18 +75,20 @@ class Interval : public DomainCreator<1> {
-  struct UpperBoundaryCondition {
-    static std::string name() noexcept { return "UpperBoundaryCondition"; }
+  struct UpperBoundary {
+    static std::string name() noexcept { return "UpperBoundary"; }
```
Is this `name` string needed, if the name matches the struct name? (also for Lower case below)

nilsvu replied:
I liked the original setup here best, where the tag was named `UpperBoundaryCondition` and its name was overridden to `UpperBoundary`. I suggest to go with that when squashing this. Up to you.

fmahebert commented on src/Domain/Creators/Interval.cpp:
```diff
@@ -27,34 +28,96 @@ struct Logical;
+  if ((is_periodic(lower_boundary_condition_) and
+       not is_periodic(upper_boundary_condition_)) or
+      (not is_periodic(lower_boundary_condition_) and
+       is_periodic(upper_boundary_condition_))) {
```
I think the shorter expression `if (is_periodic(lower) != is_periodic(upper))` has the same result? Certainly shorter to take in, though debatable if it's clearer or not... up to you if you like this or not :D

kidder replied:
I like the shorter expression

fmahebert commented on tests/Unit/Domain/Creators/Test_Interval.cpp:
```diff
@@ -83,80 +90,236 @@ void test_interval() {
+        periodic = std::make_unique<TestHelpers::domain::BoundaryConditions::
+                                        TestPeriodicBoundaryCondition<1>>();
+    const creators::Interval periodic_interval{
+        lower_bound,    upper_bound,           refinement_level[0],
```
Formatting troubles here and a few lines above

nilsdeppe replied:
If you are referring to the large spaces, this is clang-format lining up the commas. If you mean something else, I'm not sure what since running clang-format over the entire file doesn't change anything

fmahebert replied:
If this is the output of clang-format, then so be it... though I've never so strongly disagreed with it

nilsdeppe replied:
And I have yet to figure out when it decides this is the "right" formatting vs. just making a long list...

fmahebert commented:
Oops, meant to also say: fixups look good to me, you have my okay to squash them as well as any changes to my immediately above batch of comments.

nilsvu commented on src/Domain/Creators/Python/Interval.cpp:
```diff
@@ -15,9 +15,19 @@ namespace py = pybind11;
-                    std::array<bool, 1>>(),
+      .def(py::init(
+               [](std::array<double, 1> lower_x, std::array<double, 1> upper_x,
+                  std::array<bool, 1> is_periodic_in_x,
```
please move the `is_periodic_in_x` arg to the end, or move the `py::arg("is_periodic_in_x")` below to the correct position.

nilsvu commented on src/Domain/Creators/Python/Interval.cpp:
```diff
@@ -15,9 +15,19 @@ namespace py = pybind11;
+                     std::move(upper_x),
+                     std::move(initial_refinement_level_x),
+                     std::move(initial_number_of_grid_points_in_x),
+                     std::move(is_periodic_in_x),
```
clang-tidy want's you to remove the trivial `moves here

nilsvu reviewed (changes requested):
LGTM, please squash, including these

kidder reviewed (commented):
I have no further requested changes

nilsdeppe commented:
Okay, squashed and rebased. Thanks for the reviews!
